### ISSUE TITLE
Minute gets stuck in time function

### ISSUE THREAD

**Issue by tssh2005** [espressif/esp-idf#1888]
Hi everyone

I observed a strange behaviour of the `std::time` function, that occurs quite rarely. I use to following function to plot a timestamp into a logfile:

```
  char lBuffer[10U];
  std::time_t lNow = std::time(nullptr);
  struct tm* lNowTM = std::localtime(&lNow);
  std::strftime(lBuffer, sizeof(lBuffer), "%T", lNowTM);
```

It happens very rarely, that the minute value gets stuck! It simply remains on a certain value, whereas the seconds and the hours keep counting:

```
I (08:33:29) 
I (08:33:35) 2018-04-26T08:33:35+0200 - free heap: 64932 (int), 2830948 (ext)
I (08:33:35) 
I (08:33:41) 2018-04-26T08:33:41+0200 - free heap: 64932 (int), 2834228 (ext)
I (08:33:41) 
I (08:32:54) 2018-04-26T08:32:54+0200 - free heap: 64932 (int), 2834228 (ext)
I (08:32:54) 
I (08:33:00) 2018-04-26T08:33:00+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:00) 
I (08:33:07) 2018-04-26T08:33:07+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:07) 
I (08:33:13) 2018-04-26T08:33:13+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:13) 
I (08:33:19) 2018-04-26T08:33:19+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:19) 
I (08:33:26) 2018-04-26T08:33:26+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:26) 
I (08:33:32) 2018-04-26T08:33:32+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:32) 
I (08:33:38) 2018-04-26T08:33:38+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:38) 
I (08:32:51) 2018-04-26T08:32:51+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:32:51) 
I (08:32:57) 2018-04-26T08:32:57+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:32:57) 
I (08:33:03) 2018-04-26T08:33:03+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:03) 
I (08:33:10) 2018-04-26T08:33:10+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:10) 
I (08:33:16) 2018-04-26T08:33:16+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:16) 
I (08:33:22) 2018-04-26T08:33:22+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:22) 
I (08:33:29) 2018-04-26T08:33:29+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:29) 
I (08:33:35) 2018-04-26T08:33:35+0200 - free heap: 65020 (int), 2834228 (ext)
I (08:33:35) 
```
It already happend on 3 devices, which were running continuously for about a month. After a restart, everything works fine again. The time is synchronized using the SNTP-Client.
Has someone observed something similar?

**Comment by igrr** [comment]
Hi @tssh2005, could you please update the issue description to include the information requested in the issue template? Such as, IDF commit ID, what hardware is used, sdkconfig, some description of what happens in the source code (whether are you using deep sleep or light sleep, how are you initializing SNTP, and so on).

**Comment by tssh2005** [comment]
Hi, thanks for your reply!
I'm using the pre-release [v3.0-rc1](https://github.com/espressif/esp-idf/commit/043ef32651fada9f66362949deecfe101c636e6c) and ESP-WROVER hardware with 8MB flash and 4MB RAM. The system is a webserver, running continuously and waiting for incoming requests. No deep sleep or light sleep.

SNTP inititialization:
```
bool CEsp32NTPService::start()
{
  if (getenv((char*)"TZ") == NULL)
  {
    setTimezone(TIME_DEFAULT_TIMEZONE_POSIX);
  }
  initializeSntp();
  return (true);
}

void CEsp32NTPService::initializeSntp()
{
  if (sntp_enabled() == true)
  {
    // re-initialize SNTP
    sntp_stop();
  }
  strncpy(smServerName, SNTP_SERVER_NAME, SNTP_SERVER_MAX_LENGTH);
  smServerName[SNTP_SERVER_MAX_LENGTH] = '\0';
  sntp_setoperatingmode(SNTP_OPMODE_POLL);
  sntp_setservername(0, smServerName);
  sntp_init();
}
```

**Comment by tssh2005** [comment]
Interestingly, if the minute is "stuck" for hours, the minute jumps after the following message:

```
W (7197640) wifi: wait next bcn cnt == 100, beacon_interval: 104448, dtim_period: 1, next_beacon_time: 1717627155, now :2891653
```

When does this message appear? Is it possible that the wifi stack has an impact on the time module?

The whole excerpt of the log file:
```
I (20:57:36) 2018-04-25T20:57:36+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:57:36) 
I (20:57:42) 2018-04-25T20:57:42+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:57:42) 
I (20:57:48) 2018-04-25T20:57:48+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:57:49) 
I (20:57:55) 2018-04-25T20:57:55+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:57:55) 
I (20:58:01) 2018-04-25T20:58:01+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:58:01) 
I (20:58:07) 2018-04-25T20:58:07+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:58:07) 
I (20:58:14) 2018-04-25T20:58:14+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:58:14) 
I (20:58:20) 2018-04-25T20:58:20+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:58:20) 
I (20:58:26) 2018-04-25T20:58:26+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:58:26) 
I (20:57:39) 2018-04-25T20:57:39+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:57:39) 
I (20:57:45) 2018-04-25T20:57:45+0200 - free heap: 66388 (int), 2834092 (ext)
I (20:57:45) 
W (7197640) wifi: wait next bcn cnt == 100, beacon_interval: 104448, dtim_period: 1, next_beacon_time: 1717627155, now :2891653
I (21:32:45) 2018-04-25T21:32:45+0200 - free heap: 66296 (int), 2834092 (ext)
I (21:32:45) 
I (21:32:52) 2018-04-25T21:32:52+0200 - free heap: 66296 (int), 2834092 (ext)
I (21:32:52) 
I (21:32:58) 2018-04-25T21:32:58+0200 - free heap: 66296 (int), 2834092 (ext)
I (21:32:58) 
I (21:33:04) 2018-04-25T21:33:04+0200 - free heap: 66296 (int), 2834092 (ext)
I (21:33:04) 
I (21:33:11) 2018-04-25T21:33:11+0200 - free heap: 66296 (int), 2834092 (ext)
I (21:33:11) 
I (21:33:17) 2018-04-25T21:33:17+0200 - free heap: 66296 (int), 2834092 (ext)
I (21:33:17) 
I (21:32:30) 2018-04-25T21:32:30+0200 - free heap: 66296 (int), 2834092 (ext)
I (21:32:30) 
I (21:32:36) 2018-04-25T21:32:36+0200 - free heap: 66296 (int), 2834092 (ext)
I (21:32:36) 
I (21:32:42) 2018-04-25T21:32:42+0200 - free heap: 66296 (int), 2834092 (ext)
I (21:32:42) 
I (21:32:49) 2018-04-25T21:32:49+0200 - free heap: 66388 (int), 2834092 (ext)
I (21:32:49) 
I (21:32:55) 2018-04-25T21:32:55+0200 - free heap: 66388 (int), 2834092 (ext)
I (21:32:55) 
```

**Comment by tssh2005** [comment]
What does std::time use as timebase? I've found that newlib/time.c takes esp_timer as timebase. And I found that there are serveral bugs fixed in esp_timer_esp32.c that have to do with an overflow in the timer: https://github.com/espressif/esp-idf/commits/master/components/esp32/esp_timer_esp32.c

Is it possible that my problem is related to those bugfixes? Unfortunately the error is very hard reproduce, so I can not say for sure if using the fixed timer solves my problem.
